arangod/Aql/AqlValue.cpp

```diff
@@ -1674,6 +1675,27 @@ void AqlValue::initFromSlice(arangodb::velocypack::Slice slice,
                                : AqlValueType::VPACK_INLINE_INT64);
         memcpy(_data.longNumberMeta.data.slice.slice, slice.begin(),
                static_cast<size_t>(length));
+        // If length == 9, we're done;
+        // If length == 8, there's one byte left to fill.
+
+        // For signed ints, do an *arithmetic* shift of the biggest byte. So
+        // we get 0xff for negative, and 0x00 for all nonnegative integers.
+        auto const filler =
+            slice.isUInt()
+                ? 0x00
+                : std::int8_t(slice.begin()[length - 1]) >> (CHAR_BIT - 1);
+        auto const remainingBytes =
+            sizeof(_data.longNumberMeta.data.slice.slice) - length;
+        // Set the remaining byte (if any):
+        std::memset(_data.longNumberMeta.data.slice.slice + length, filler,
+                    remainingBytes);
+        TRI_ASSERT(
+            slice.isUInt()
+                ? slice.getUInt() > std::numeric_limits<std::int64_t>().max() ||
+                      slice.getUInt() == static_cast<uint64_t>(this->toInt64())
+                : slice.getInt() == this->toInt64())
+            << (slice.isUInt() ? "uint " : "int ") << slice.toJson()
+            << " != " << this->toInt64();
```

Comment:
```suggestion
        auto& last_byte = _data.longNumberMeta.data.slice.slice[8];
        last_byte = 0x00;
        memcpy(_data.longNumberMeta.data.slice.slice, slice.begin(),
               static_cast<size_t>(length));
        // If length == 9, we're done;
        // If length == 8, there's one byte left to fill if it's signed int;
        if (length == 8 && !slice.isUint() && slice.begin()[7] > 0x7F) {
          last_byte = 0xFF;
        }
        TRI_ASSERT(
            slice.isUInt()
                ? slice.getUInt() > std::numeric_limits<std::int64_t>().max() ||
                      slice.getUInt() == static_cast<uint64_t>(this->toInt64())
                : slice.getInt() == this->toInt64())
            << (slice.isUInt() ? "uint " : "int ") << slice.toJson()
            << " != " << this->toInt64();
```

signed shift for negative values is implementation defined behavior.
Also any arguments smaller than int implicitly casted to the int.
why not simplify it?

Also it looks like `!slice.isUint()` in code above unnecessary.

Reply:
Actually, this was fixed in C++20, so signed shifts are now well defined.

Reply:
Good, but still calling memset with runtime size doesn't look good for me

Reply:
> Also any arguments smaller than int implicitly casted to the int.
> why not simplify it?

It's written this way for the reader, so the intent is clear. The compiler doesn't care either way.

> Also it looks like !slice.isUint() in code above unnecessary.

I don't know what you mean. There's no `!slice.isUint()` in my PR, only in your suggestion. And there it's strictly necessary.

> Good, but still calling memset with runtime (for compiler) size doesn't look good for me

Sure, I've removed the call to memset.

Reply:
> I don't know what you mean. There's no !slice.isUint() in my PR, only in your suggestion. And there it's strictly necessary

Yeah you're right.

Reply:
> It's written this way for the reader, so the intent is clear. The compiler doesn't care either way.

Well my concerns about it's pretty tricky.

That we have int(0x00) in unsigned part of condition, and int(0x00) or int(0xFFFFFFFF), not 0xFF, in other

Reply:
I understand. I've changed it to a branch.